### PR TITLE
addEventListener() and New Events

### PR DESCRIPTION
**Pull request by vcoindet**
All changes are in ApiGlobe.js.

1 - ApiGlobe.prototype.addEventListener = function (eventname, callback){

```
this.viewerDiv.addEventListener(eventname, callback, false);
```

};

2 - Adding the following Events :
centerchanged 
zoomchanged
orientationchanged
layerchanged In Progress
layerchanged:visible
layerchanged:opacity
layerchanged:index
layeradded
layerremoved 
error

3 - Dispatching the events in other functions (setCenter, setCenterAdvanced, removeImageryLayer, setLayerVisibility etc...)

Tested with ie11, chrome 53 and firefox 38.4

### CODE REVIEW

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 1]
```diff
@@ -20,8 +20,19 @@ import {STRATEGY_MIN_NETWORK_TRAFFIC} from 'Scene/LayerUpdateStrategy'
 
 var loaded = false;
 var eventLoaded = new CustomEvent('globe-loaded');
-var eventLayerRemoved = new CustomEvent('Layer-removed');
+//var eventLayerRemoved = new CustomEvent('Layer-removed');
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -114,6 +126,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
 
 ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
     this.scene.getMap().layersConfiguration.moveLayerToIndex(layer, newId);
+    eventLayerChanged.layerId = newId;
+    eventLayerChangedIndex.layerId = newId;
+    //eventLayerChanged.initCustomEvent("layerchanged:index", false, false, false);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 3]
```diff
@@ -212,6 +231,8 @@ ApiGlobe.prototype.createSceneGlobe = function(coordCarto, viewerDiv) {
 
             loaded = true;
             viewerDiv.dispatchEvent(eventLoaded);
+        }else{
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Change format

``` js
}else{
```

to

``` js
} else {
```

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 4]
```diff
@@ -306,8 +327,13 @@ ApiGlobe.prototype.setRealisticLightingOn = function(value) {
 ApiGlobe.prototype.setLayerVisibility = function(id, visible) {
 
     this.scene.getMap().setLayerVisibility(id, visible);
-
+    eventLayerChanged.layerId = id;
+    eventLayerChangedVisible.layerId = id;
+    //eventLayerChanged.initCustomEvent("layerchanged:visible", false, false, false);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 5]
```diff
@@ -306,8 +327,13 @@ ApiGlobe.prototype.setRealisticLightingOn = function(value) {
 ApiGlobe.prototype.setLayerVisibility = function(id, visible) {
 
     this.scene.getMap().setLayerVisibility(id, visible);
-
+    eventLayerChanged.layerId = id;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Need Factorization 

Listen all Layers changes (visibilty, opacity, id...) and `dispatchEvent(eventLayerChanged)`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -330,6 +356,11 @@ ApiGlobe.prototype.orbit = function(value) {
 ApiGlobe.prototype.setLayerOpacity = function(id, visible) {
 
     this.scene.getMap().setLayerOpacity(id, visible);
+    eventLayerChanged.layerId = id;
+    eventLayerChangedOpacity.layerId = id;
+    //eventLayerChanged.initCustomEvent("layerchanged:opacity", false, false, false);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -470,8 +500,9 @@ ApiGlobe.prototype.getRange = function() {
  */
 
 ApiGlobe.prototype.setTilt = function(tilt /*, bool*/ ) {
-
+    eventOrientation.oldTilt = this.getTilt();
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:494]: Why old tilt?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 8]
```diff
@@ -483,8 +514,9 @@ ApiGlobe.prototype.setTilt = function(tilt /*, bool*/ ) {
  */
 
 ApiGlobe.prototype.setHeading = function(heading /*, bool*/ ) {
-
+    eventOrientation.oldHeading = this.getHeading();
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:508]: Why old Heading?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 9]
```diff
@@ -529,8 +561,10 @@ ApiGlobe.prototype.computeDistance = function(p1, p2) {
  */
 
 ApiGlobe.prototype.setCenter = function(coordinates) {
+    eventCenter.oldCenter = this.getCenter();
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:555]: Why old center?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 10]
```diff
@@ -549,6 +583,9 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
     this.setRange(pPosition.range);
     this.setHeading(pPosition.heading);
     this.setTilt(pPosition.tilt);
+    /*this.viewerDiv.dispatchEvent(eventRange);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove dead code

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 11]
```diff
@@ -560,10 +597,10 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
  */
 
 ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
-    var viewerDiv = document.getElementById("viewerDiv");
-
+    eventRange.oldRange = this.getRange();
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:588]: why old range?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 12]
```diff
@@ -560,10 +597,10 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
...
     this.scene.currentControls().setRange(pRange);
-    viewerDiv.dispatchEvent(eventRange);
+    this.viewerDiv.dispatchEvent(eventRange);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:590]: Need to centralize events -> one event by function

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 13]
```diff
@@ -114,6 +126,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
 
 ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
     this.scene.getMap().layersConfiguration.moveLayerToIndex(layer, newId);
+    eventLayerChanged.layerId = newId;
+    eventLayerChangedIndex.layerId = newId;
+    //eventLayerChanged.initCustomEvent("layerchanged:index", false, false, false);
+    //var eventLayerChangedVisible = new CustomEvent('layerchanged:visible');
+    this.viewerDiv.dispatchEvent(eventLayerChanged);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Centralize event -> one event by function

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 14]
```diff
@@ -306,8 +327,13 @@ ApiGlobe.prototype.setRealisticLightingOn = function(value) {
 ApiGlobe.prototype.setLayerVisibility = function(id, visible) {
 
     this.scene.getMap().setLayerVisibility(id, visible);
-
+    eventLayerChanged.layerId = id;
+    eventLayerChangedVisible.layerId = id;
+    //eventLayerChanged.initCustomEvent("layerchanged:visible", false, false, false);
+    this.viewerDiv.dispatchEvent(eventLayerChanged);
+    this.viewerDiv.dispatchEvent(eventLayerChangedVisible);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:328]: Centralize event -> one event by function

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 15]
```diff
@@ -330,6 +356,11 @@ ApiGlobe.prototype.orbit = function(value) {
 ApiGlobe.prototype.setLayerOpacity = function(id, visible) {
 
     this.scene.getMap().setLayerOpacity(id, visible);
+    eventLayerChanged.layerId = id;
+    eventLayerChangedOpacity.layerId = id;
+    //eventLayerChanged.initCustomEvent("layerchanged:opacity", false, false, false);
+    this.viewerDiv.dispatchEvent(eventLayerChanged);
+    this.viewerDiv.dispatchEvent(eventLayerChangedOpacity);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:355]: Centralize event -> one event by function

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 16]
```diff
@@ -576,6 +613,12 @@ ApiGlobe.prototype.getZoomLevel = function(id) {
     return this.scene.getMap().getZoomLevel(id);
 };
 
+ApiGlobe.prototype.addEventListener = function (eventname, callback){
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: replace by Api function : ` itowns.viewer.addEventListener`

File: index.html [hunk 17]
```diff
@@ -694,9 +694,9 @@
 
         viewerDiv.addEventListener('globe-loaded', function() {
```
Comment (gchoqueux) [inline index.html]: replace by Api function : `itowns.viewer.addEventListener`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 4 above)
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Called when the visibility of a layer has changed, return an object with two properties, id and visible

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 18]
```diff
@@ -330,6 +356,11 @@ ApiGlobe.prototype.orbit = function(value) {
 ApiGlobe.prototype.setLayerOpacity = function(id, visible) {
 
     this.scene.getMap().setLayerOpacity(id, visible);
+    eventLayerChanged.layerId = id;
+    eventLayerChangedOpacity.layerId = id;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:353]: Called when the opacity of a layer has changed, return an object with two properties, id and opacity

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 19]
```diff
@@ -114,6 +126,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
 
 ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
     this.scene.getMap().layersConfiguration.moveLayerToIndex(layer, newId);
+    eventLayerChanged.layerId = newId;
+    eventLayerChangedIndex.layerId = newId;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Called when the index of a layer has changed, return an object with two properties, id and index

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:494]: @gchoqueux : It was a request from the IGN's Geoportail.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 8 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:508]: @gchoqueux : It was a request from the IGN's Geoportail.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 9 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:555]: @gchoqueux : It was a request from the IGN's Geoportail.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 11 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:588]: @gchoqueux : It was a request from the IGN's Geoportail.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 20]
```diff
@@ -114,6 +124,9 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
 
 ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
     this.scene.getMap().layersConfiguration.moveLayerToIndex(layer, newId);
+    eventLayerChangedIndex.layerId = newId;
+    eventLayerChangedIndex.layer = layer;
+    this.viewerDiv.dispatchEvent(eventLayerChangedIndex);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:131]: place after  `this.scene.getMap().updateLayersOrdering();`

**Pull request merged**